ARCs/arc-0060.md

```diff
@@ -0,0 +1,506 @@
+
+Algorand wallets need a standard approach to byte signing to unlock self-custodial services and protect users from malicious and attack-prone signing workflows.
+
+This ARC provides a standard API for bytes signing. The API encodes byte arrays into well structired JSON schemas together with additional metadata. It requires wallets to validate the signing inputs, notify users about what they are signing and warn them in case of dangerous signing requests.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative
+
+### Overview
+
+> This section is non-normative
+
+Signatures of bytes are processed with the `signData(arbData, metadata)` function. It requires a non-empty list of structured `arbData` and `metadata` objects.
+
+A `StdSignData` object represents structured arbitrary data that must be signed. Those data include a field `data` that is an array of bytes and a field `signers` for the signing keys.
+
+A `StdSignMetadata` object describes the signature scope and the type of data being signed, including their encoding and the JSON schema.
+
+`data` is structured according to a JSON Schema provided with the metadata.
+
+There are two possible use cases:
+
+1. Sign data with one signer that the wallet knows. For example:
+
+```json
+{
+  "data": "/{.../}",
+  "signers": ["xxxx"]
+}
+{
+  "scope": "ARBITRARY",
+  "schema": "...",
+  "message": "These are just arbitrary bytes",
+  "encoding": "..."
+}
+```
+
+2. Sign data with a secondary public key that the wallet knows (recalling the Algorand Rekey) or a multisig address derived by two or more public keys that the wallet knows (recalling the Algorand MultiSig), or a hierarchical deterministic (HD) wallet. For example:
+
+```json
+{
+  "data": "/{.../}",
+  "signers": ...,
+  "authAddr": "xxxx",
+  "msig": {
+    "version": 1,
+    "threshold": 2,
+    "addrs": [
+      "xxxxxx",
+      "xxxxxx",
+      "xxxxxx"
+    ]
+  },
+  "hdPath": {
+    "purpose": 44,
+    "coinType": 0,
+    "account": 0,
+    "change": 0,
+    "addrIdx": 0
+  }
+}
+{
+  "scope": "ARBITRARY",
+  "schema": "...",
+  "message": "These are just arbitrary bytes",
+  "encoding": "..."
+}
+```
+
+### Interfaces
+
+> Interfaces are defined in TypeScript. All the objects that are defined are valid JSON objects.
+
+ARC-60 uses interchangeably the terms "throw an error" and "reject a promise with an error".
+
+#### **Interface `SignDataFunction`**
+
+A wallet function `signData` that signs arbitrary data is defined by the interface:
+
+```tsx
+export type SignDataFunction = {
+    arbData: StdSignData[],
```

Comment:
Why does it need to bulk sign? Most Key Management Systems do not support bulk signing and i can't see the purpose of doing it here. 

If the client needs multiple signatures, if it should be done multiple times. So it's also authorized, checked, logged multiple times

Reply:
Good point. Although I think it is more friendly for users to have bulk signs (e.g. signing documents), I also agree that we should mimic the state-of-the-art signing approaches like KMS/HSMs. I will change it to single signing.

Reply:
https://github.com/algorandfoundation/ARCs/pull/284/commits/67754a89ff9e211cfee6f29ebd7aa4f016b19243